### DISCUSSION TITLE
Faust MIDI or Faust plus Arduino-MIDI?

### DISCUSSION THREAD

**Discussion by jonrichings** [pschatzmann/arduino-audio-tools#182] [General]
Since Faust supports it's own MIDI solution I'm curious as to thoughts of which MIDI mechanism is best use.  My gut would prefer Phil's solution as I'd like to keep all the sound management and routing in Audio-Tools.  But I wonder if there is any pros/cons to be weighed up.  

I'm also struggling to get my Arduino-BLE-MIDI input to control Faust, both controllers and key input.  

And I think Faust's MIDI is expecting a higher-level MIDI interface from hosts with an OS.  I can see no way to tell Faust where to find the MIDI input.

**Comment by sletz** [comment]
See: https://faustdoc.grame.fr/manual/architectures/#midi-architecture-modules and possibly https://github.com/grame-cncm/faust/blob/master-dev/architecture/faust/midi/gramophone-midi.h

**Reply by jonrichings** [reply]
That's helpful, thanks.  I hadn't seen the gramophone one.

**Comment by jonrichings** [comment]
Still working on this.  I can get BLE-MIDI working with audio tools and variations of the examples.  But I am still wondering if using Faust's built-in MIDI architecture wouldn't be preferable.  The Faust MIDI architecture is quite neat and flexible but lacks BLE as an input.  I can't get my head around how Gramophone implements BLE-MIDI.  What I think I need is a way to take BLE-MIDI from Phil's lib and pass it to Faust's MIDI handler.  Anybody care to suggest how that would work?

**Reply by sletz** [reply]
The general idea is to write a subclass of [midi_handller](https://github.com/grame-cncm/faust/blob/master-dev/architecture/faust/midi/midi.h#L288), implement the `startMidi/stopMidi` methods and a MIDI message decoding callback. Looking at already implemented classes in https://github.com/grame-cncm/faust/tree/master-dev/architecture/faust/midi can help.

**Reply by jonrichings** [reply]
Thanks Stéphane, 

I'm afraid I'm a hardware guy with only old-skool c skills.  I've never understood OOP stuff.  

If I just included gramophone-midi.h, added startMidi and stopMidi in appropriate places, then added a callback_midi_message_received() function to handle the received ble-midi, would that be enough?

**Reply by sletz** [reply]
`callback_midi_message_received` is setup in `blemidi_init` here https://github.com/grame-cncm/faust/blob/master-dev/architecture/faust/midi/gramophone-midi.h#L89 and called by the hardware.

Yes `startMidi` and `stopMidi` have to be used at appropriate places.

**Reply by jonrichings** [reply]
Thanks Stéphane,  I've been trying to get this work but all the dependent includes are tripping me up.  Going back to a simpler approach based on Phil's BLE-Midi and trying to talk to the Faust DSP.
